### ISSUE TITLE
Customer_churn_prediction

### ISSUE THREAD

**Issue by amrutadpatil2001-droid** [amrutadpatil2001-droid/Python_Projects#13]
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler, MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

#Load dataset
# Load Dataset
df = pd.read_csv("customer_churn_data.csv")
print(df.head())
print(df.info())

#Data cleaning 
# Lowercasing the coulmn names
df.columns = df.columns.str.lower()
df

# Basic Cleaning
df.drop('customerid', axis=1, inplace=True)

# Convert TotalCharges to numeric
df['totalcharges'] = pd.to_numeric(df['totalcharges'], errors='coerce')
df.dropna(inplace=True)

#Encoding

# Encode Target Variable (Label Encoding)
le = LabelEncoder()
df['churn'] = le.fit_transform(df['churn'])  # Yes=1, No=0

#Feature Engineering (5+ features)

# 1. Avg Monthly Spend
df['avgmonthlyspend'] = df['totalcharges'] / (df['tenure'] + 1)

# 2. Tenure Group
df['tenuregroup'] = pd.cut(df['tenure'], bins=[0,12,24,48,72], labels=[0,1,2,3])

# 3. High Charges Flag
df['highcharges'] = (df['monthlycharges'] > df['monthlycharges'].median()).astype(int)

# 4. Long Term Customer
df['longterm'] = (df['tenure'] > 24).astype(int)

# 5. Charges per Tenure
df['chargespertenure'] = df['totalcharges'] / (df['tenure'] + 1)

df[['avgmonthlyspend','chargespertenure','highcharges','longterm']].head(10)

#Handle Outliers (IQR Method)
Q1 = df['monthlycharges'].quantile(0.25)
Q3 = df['monthlycharges'].quantile(0.75)
IQR = Q3 - Q1

df = df[~((df['monthlycharges'] < (Q1 - 1.5 * IQR)) |
          (df['monthlycharges'] > (Q3 + 1.5 * IQR)))]

#Split Features & Target
X = df.drop('churn', axis=1)
y = df['churn']

#Identify Columns
num_cols = X.select_dtypes(include=['int64', 'float64']).columns
cat_cols = X.select_dtypes(include=['object', 'category']).columns

#Preprocessing Pipeline

preprocessor = ColumnTransformer([
    ('num_standard', StandardScaler(), num_cols),   # Scaling 1
    ('cat_onehot', OneHotEncoder(drop='first'), cat_cols)  # Encoding 2
])

# Model Pipeline
pipeline = Pipeline([
    ('preprocessing', preprocessor),
    ('model', RandomForestClassifier(n_estimators=100, random_state=42))
])

# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Train Model
pipeline.fit(X_train, y_train)

# Predictions
y_pred = pipeline.predict(X_test)

# Evaluation
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
